routes/index.js

```diff
@@ -1,367 +1,376 @@
-// // In order of simplicity we are not using any database. But you can write the
-// // same logic using MongoDB.
-// const users = [
-//   // You know password for the user.
-//   { name: 'user', password: 'pwd' },
-//   // You don't know password for the admin.
-//   { name: 'admin', password: Math.random().toString(32), canDelete: true },
-// ];
-
-// let messages = [];
-// let lastId = 1;
-
-// function findUser(auth) {
-//   return users.find((u) =>
-//     u.name === auth.name &&
-//     u.password === auth.password);
-// }
-// ///////////////////////////////////////////////////////////////////////////////
-
-// exports.chat = {
-//   get(req, res) {
-//     res.send(messages);
-//   },
-//   add(req, res) {
-//     const user = findUser(req.body.auth || {});
-
-//     if (!user) {
-//       res.status(403).send({ ok: false, error: 'Access denied' });
-//       return;
-//     }
-
-//     const message = {
-//       // Default message icon. Cen be overwritten by user.
-//       icon: '👋',
-//     };
-
-//     _.merge(message, req.body.message, {
-//       id: lastId++,
-//       timestamp: Date.now(),
-//       userName: user.name,
-//     });
-
-//     messages.push(message);
-//     res.send({ ok: true });
-//   },
-//   delete(req, res) {
-//     const user = findUser(req.body.auth || {});
-
-//     if (!user || !user.canDelete) {
-//       res.status(403).send({ ok: false, error: 'Access denied' });
-//       return;
-//     }
-
-//     messages = messages.filter((m) => m.id !== req.body.messageId);
-//     res.send({ ok: true });
-//   }
-// };
+var utils = require('../utils');
+var mongoose = require('mongoose');
+var Todo = mongoose.model('Todo');
+var User = mongoose.model('User');
+// TODO:
+var hms = require('humanize-ms');
+var ms = require('ms');
+var streamBuffers = require('stream-buffers');
+var readline = require('readline');
+var moment = require('moment');
+var execFile = require('child_process').execFile;
+var RateLimit = require('express-rate-limit');
+var limiter = new RateLimit({
+  windowMs: parseInt(process.env.WINDOW_MS, 10),
+  max: parseInt(process.env.MAX_IP_REQUESTS, 10),
+  delayMs:parseInt(process.env.DELAY_MS, 10),
+  headers: true
+});
+
+app.use(limiter);
+var validator = require('validator');
+
+// zip-slip
+var fileType = require('file-type');
+var AdmZip = require('adm-zip');
+var fs = require('fs');
+
+// prototype-pollution
+var _ = require('lodash');
+
+exports.index = function (req, res, next) {
+  Todo.
+    find({}).
+    sort('-updated_at').
+    exec(function (err, todos) {
+      if (err) return next(err);
+
+      res.render('index', {
+        title: 'Patch TODO List',
+        subhead: 'Vulnerabilities at their best',
+        todos: todos,
+      });
+    });
+};
+
+exports.loginHandler = function (req, res, next) {
+  if (validator.isEmail(req.body.username)) {
+    User.find({ username: req.body.username, password: req.body.password }, function (err, users) {
```

Comment:
### <sub><img src="https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg" width="20"/></sub>&nbsp; NoSQL Injection

Unsanitized input from the HTTP request body flows into find, where it is used in an NoSQL query. This may result in an NoSQL Injection vulnerability.

<h6>
<a href="https://cwe.mitre.org/data/definitions/943.html">CWE-943</a> | <a href="https://docs.snyk.io/manage-risk/prioritize-issues-for-fixing/priority-score">Priority score 860</a> | Line 48</h6>